### PR TITLE
Enable basic support for `extension.worktreeconfig` with `core.repositoryformatversion=1`  to unblock sparse checkout in partial clones.

### LINKED ISSUE #195: [Upstream Issue] Libgit2 and transitively Stacked Git do not support sparse checkouts
Upstream Issues:
* https://github.com/libgit2/libgit2/issues/6044
* https://github.com/libgit2/libgit2/issues/2263

Adding this issue to track upstream issues of LibGit2 related to sparse checkout support.

Our company is using a huge monorepo, and we have recently onboarded onto Git Sparse checkouts, which seem to be unsupported in LibGit2 which Rust version is using.

Despite my attempts at doing custom builds with some PRs I found online for LibGit2, although I could get the Read Only workflow to succeed like `stg series`.

I could not get the `stg push` to work as it seems like sparse index is throwing LibGit2 way off.

### PR DESCRIPTION
**Pull request by NonLogicalDev**
`libgit2` library by default fails any operation if it encounters any unrecognised extensions configured in a repo, as per Git documentation.

ref: https://git-scm.com/docs/repository-version

> If a version-1 repository specifies any extensions.* keys that the
> running git has not implemented, the operation MUST NOT proceed.
> Similarly, if the value of any known key is not understood by the
> implementation, the operation MUST NOT proceed.

If a repository is using a sparse checkout mode it force enables the `worktreeconfig` extensions which allows setting configuration in repo on a per worktree basis, to enable some of the sparse checkout functionality.

Since `stg` is using `libgit2` library only as an interface to object database and for looking up alias lookup from git configuration (i.e. Read Only operation) and uses git cli for everything else, it is relatively safe to ignore per worktree git configuration.

This patch adds `worktreeconfig` extension to the `libgit2` extension whitelist.

CLOSES #195

### CODE REVIEW

**Review by jpgrayson — CHANGES_REQUESTED** [review]
One thing I still don't understand is how the tests in `t7000-sparse-checkout.sh` pass without this change? It seems like as long as an associated worktree is not added with `git worktree add`, then libgit2 is okay with `extensions.worktreeConfig` being in `.git/config`.

It would be helpful if there was a test, either in `t7000-sparse-checkout.sh` or in a new test script that triggers the problem that this change solves.

File: Cargo.toml [hunk 1]
```diff
@@ -40,13 +40,15 @@ strsim = "0.10"
 tempfile = "3"
 termcolor = "1.1"
 thiserror = "~1.0"
+once_cell = "~1.14"
```
[the remark this answers is not in the document]
Comment (NonLogicalDev) [reply Cargo.toml:43]: fair point will take this out, this was leftover part of my experiment to use a singleton based initialization.
But I was sadly incapable of figuring out a way to code it that appeased the borrow checker.

File: Cargo.toml [hunk 2]
```diff
@@ -40,13 +40,15 @@ strsim = "0.10"
...
 curl = { version = "0.4", optional = true }
 flate2 = { version = "1", optional = true }
 tar = { version = "0.4", optional = true }
 
 [features]
-default = ["import-compressed", "import-url"]
+default = ["import-compressed", "import-url", "git2-ext-worktreeconfig-support"]
 import-compressed = ["dep:bzip2", "dep:flate2", "dep:tar"]
 import-url = ["dep:curl"]
+git2-ext-worktreeconfig-support = []
```
[the remark this answers is not in the document]
Comment (NonLogicalDev) [reply Cargo.toml:54]: I was thinking of keeping it there for safety. I suppose a better way might be to have it be guarded with a config flag.

File: src/main.rs [hunk 3]
```diff
@@ -138,6 +166,9 @@ pub(crate) fn get_full_command(
 /// The name of the game is to dispatch to the appropriate subcommand or alias as
 /// quickly as possible.
 fn main() -> ! {
+    // Initialize git2 library.
+    unsafe { init_libgit2() }.expect("Failed libgit2 init");
+
```
[the remark this answers is not in the document]
Comment (NonLogicalDev) [reply src/main.rs:171]: That sounds good to me. I did not notice the slowdown but I get where you are coming from.
Ideally this should be something of a global helper singleton function, called upon needing to access the repo for first time.

**Comment by NonLogicalDev** [comment]
@jpgrayson I figured out why `t7000-sparse-checkout.sh` works and why our company monorepo does not.

It all has to do with `core.repositoryversion`. I dug in and found out that `core.repositoryversion` gets set to 1 whey the Object Hash is configured to be SHA256 rather than SHA1.

I threw together a test script that reproduces this issue:
https://gist.github.com/NonLogicalDev/29b55d855c8aaae0a641cd0dda2ec443

**Review by jpgrayson — APPROVED** [review]

**Comment by jpgrayson** [comment]
Thank you for solving the mystery of the difference between `t7000-sparse-checkout.sh` and your environment! And for the test case. I will do a follow-on change to integrate the test case into StGit's test suite.

**Comment by NonLogicalDev** [comment]
Huge Thanks @jpgrayson! Super grateful for your support and patience on this project!

**Comment by jpgrayson** [comment]
Thank you @NonLogicalDev for doing all the legwork to figure out this problem.

**Pull request merged**